python/smqtk/web/nn_descriptor_service/__init__.py

```diff
@@ -0,0 +1,230 @@
+            json_config['nn_index'],
+            get_nn_index_impls
+        )
+
+        self.descriptor_generator_inst = plugin.from_plugin_config(
+                                            self.generator_config,
+                                            get_descriptor_generator_impls)
+
+
+        @self.route("/nn/<path:uri>")
+        def compute_nearest_neighbors(uri):
+            """
+            Data modes for upload/use::
+
+                - local filepath
+                - base64
+                - http/s URL
+
+            The following sub-sections detail how different URI's can be used.
+
+            Local Filepath
+            --------------
+
+            The URI string must be prefixed with ``file://``, followed by the
+            full path to the data file to describe.
+
+            Base 64 data
+            ------------
+
+            The URI string must be prefixed with "base64://", followed by the
+            base64 encoded string. This mode also requires an additional
+            ``?content_type=`` to provide data content type information. This
+            mode saves the encoded data to temporary file for processing.
+
+            HTTP/S address
+            --------------
+
+            This is the default mode when the URI prefix is none of the above.
+            This uses the requests module to locally download a data file
+            for processing.
+
+            JSON Return format::
+                {
+                    "success": <bool>
+
+                    "message": <str>
+
+                    "neighbors": <None|list[float]>
+
+                    "reference_uri": <str>
+                }
+
+            :type descriptor_label: str
+            :type uri: str
+
+            """
+            message = "execution nominal"
+            descriptor = None
+
+            de = None
+            try:
+                de = self.resolve_data_element(uri)
+            except ValueError, ex:
+                message = "URI resolution issue: %s" % str(ex)
+
+            if de:
+                try:
+                    descriptor = self.descriptor_generator_inst.\
+                        compute_descriptor(de, self.descr_elem_factory)
+
+                    if descriptor is not None:
+                        descriptor.set_vector(descriptor.vector().flatten())
+                    else
+                        self.log.error("Descriptor is null or invalid")
+                except RuntimeError, ex:
+                    message = "Descriptor extraction failure: %s" % str(ex)
+                except ValueError, ex:
+                    message = "Data content type issue: %s" % str(ex)
+
+            # fail here if de is None
+            # Default is 8
+            num_neighbors = flask.request.args.get("num_neighbors", 8)
+
+            neighbors = []
+            if descriptor is not None:
+                neighbors, _ = self.nn_index.nn(descriptor, n=num_neighbors)
```

Comment:
This could raise ``ValueError`` exceptions (if descriptor didn't have a vector, or if the index didn't contain anything), so we're going to want to protect against that.